engines/dradis-api/lib/dradis/ce/api/basic_auth_strategy.rb

```diff
@@ -1,13 +1,5 @@
-# Dradis::API::WardenStrategy
-#
-# HTTP Basic authentication strategy for Warden.
-#
-# See:
-#   https://github.com/hassox/warden
-#
 module Dradis::CE::API
-  class WardenStrategy < ::Warden::Strategies::Base
-
+  class BasicAuthStrategy < ::Warden::Strategies::Base
```

Comment:
this strategy quacks a lot like the [:shared_password](https://github.com/dradis/dradis-ce/blob/3377d759a15008e577be9cfb0af9938034415910/config/initializers/warden_00_shared_password.rb) one.

I think it's worth creating a `./lib/dradis/ce/warden/shared_password` strategy that has an `authenticate!` method that uses the `username` and `password` values to perform the secure check (removing the need to have `User.authenticate` as mentioned above). In the `SharedPassword` strategy the `username` method can take the value from the `params` and in the API streategy we can override the method to take the value from `auth.credentials`.

That way we end up with 1 place in the entire codebase where we're doing the `BCrypt` check for the shared password.

If you think that's too much (I could buy that argument particularly because the "fail" condition in both strategies isn't standard. Then I'd also buy two separate strategies classes in which the `authenticate!` method is pretty much identical (relies on `username`, `password`, and does the `BCrypt` check. Then we move the implementation of where the username comes from outside the `authenticate!` method to a separate `def username` method.